types/jest/index.d.ts

```diff
@@ -1148,6 +1149,10 @@ declare namespace jest {
          * to restore mocks automatically between tests.
          */
         mockRestore(): void;
+        /**
+         * Returns the function that was set as the implementation of the mock (using mockImplementation).
+         */
+        getMockImplementation(): (...args: Y) => T | undefined;
```

Comment:
This definition is wrong. According to https://github.com/facebook/jest/blob/master/packages/jest-mock/src/index.ts#L51, this should be `getMockImplementation(): ((...args: Y) => T) | undefined;`. I think you just missed a couple of parenthesis.

Reply:
@GerkinDev I wouldn't trust all the type definitions you see on jest site.

Reply:
@tonyhallett this isn't from jest site, but from source code type definitions. Moreover, in real use cases, if actual implementation is not defined, it returns `undefined`, not a function with same arity that may return undefined.

Anyway, I've PR: #50377

Reply:
@GerkinDev I was referring to the jest source code type definitions.  My comment was in case you were attempting to match theirs without reading the source code.

> According to https://github.com/facebook/jest/blob/master/packages/jest-mock/src/index.ts#L51

Although the jest type definition has getMockImplementation(): Function | undefined;


But
https://github.com/facebook/jest/blob/7430a7824421c122cd07035d800d22e1007408fa/packages/jest-mock/src/index.ts#L622

```
f.getMockImplementation = () => this._ensureMockConfig(f).mockImpl;
```

Also see https://github.com/facebook/jest/issues/9061

> Regarding getMockImplementation, I think that's undocumented and should not be used


So your change to 
> getMockImplementation(): ((...args: Y) => T) | undefined;
in #50377 seems incorrect to me.

Reply:
Given

> ```ts
> f.getMockImplementation = () => this._ensureMockConfig(f).mockImpl;
> ```
> https://github.com/facebook/jest/blob/7430a7824421c122cd07035d800d22e1007408fa/packages/jest-mock/src/index.ts#L622

This calls 

> ```ts
> private _ensureMockConfig<T, Y extends Array<unknown>>(
>     f: Mock<T, Y>,
>   ): MockFunctionConfig {
>     let config = this._mockConfigRegistry.get(f);
>     if (!config) {
>       config = this._defaultMockConfig();
>       this._mockConfigRegistry.set(f, config);
>     }
>     return config;
>   }
>   ```
> https://github.com/facebook/jest/blob/7430a7824421c122cd07035d800d22e1007408fa/packages/jest-mock/src/index.ts#L432-L441

If there is no config, it inits a new one with

> ```ts
>   private _defaultMockConfig(): MockFunctionConfig {
>     return {
>       mockImpl: undefined,
>       mockName: 'jest.fn()',
>       specificMockImpls: [],
>       specificReturnValues: [],
>     };
>   }
> https://github.com/facebook/jest/blob/7430a7824421c122cd07035d800d22e1007408fa/packages/jest-mock/src/index.ts#L454-L461

where `mockImpl` is `undefined`.

Back in your sample

> ```ts
> f.getMockImplementation = () => this._ensureMockConfig(f).mockImpl;
> ```
> https://github.com/facebook/jest/blob/7430a7824421c122cd07035d800d22e1007408fa/packages/jest-mock/src/index.ts#L622

it's actually the `mockImpl` property that is returned.

I maintain my PR and am even more sure of it.

Reply:
Please continue the conversation in #50377. I'll re-post this comment, and I think we should continue there.